PhysicsTools/PyTorch/plugins/alpaka/AotRegressionProducer.cc

```diff
@@ -0,0 +1,131 @@
+
+  private:
+    const device::EDGetToken<torchportable::ParticleCollection> inputs_token_;    /**< Token to get input data. */
+    const device::EDPutToken<torchportable::RegressionCollection> outputs_token_; /**< Token to store output data. */
+    std::unique_ptr<Kernels> kernels_ = nullptr; /**< Kernel utilities for post-inference validation. */
+    std::unique_ptr<AotModel> model_ = nullptr;  /**< Cache for the AOT model. */
+  };
+
+  AotRegressionProducer::AotRegressionProducer(edm::ParameterSet const &params)
+      : EDProducer<>(params),
+        inputs_token_{consumes(params.getParameter<edm::InputTag>("inputs"))},
+        outputs_token_{produces()},
+        kernels_{std::make_unique<Kernels>()} {
+    cms::torch::alpaka::set_threading_guard();
+    model_ = std::make_unique<AotModel>(params.getParameter<edm::FileInPath>("modelPath").fullPath());
+  }
+
+  /**
+   * @brief Performs inference on the event data.
+   * @param event The event to process.
+   * @param event_setup Event setup information.
+   */
+  void AotRegressionProducer::produce(device::Event &event, const device::EventSetup &event_setup) {
+    auto t1 = std::chrono::high_resolution_clock::now();
+
+    // debug stream usage in concurrently scheduled modules
+    std::stringstream msg_stream;
+    msg_stream << "RegressionAot::produce [E: " << event.id().event() << "]";
+    auto msg = msg_stream.str();
+    NvtxScopedRange produce_range(msg.c_str());
+
+    // guard torch internal operations to not conflict with cmssw fw scheme
+    cms::torch::alpaka::Guard<Queue> guard(event.queue());
+    // sanity check for debug
+    assert(cms::torch::alpaka::queue_hash(event.queue()) == cms::torch::alpaka::current_stream_hash(event.queue()));
+
+    // get data
+    // TODO: const_cast should not be done by user
+    // in principle should not be done by anyone
+    // @see: torch::from_blob(void*)
+    auto &inputs = const_cast<torchportable::ParticleCollection &>(event.get(inputs_token_));
+    const size_t batch_size = inputs.const_view().metadata().size();
+    auto outputs = torchportable::RegressionCollection(batch_size, event.queue());
+
+    // metadata for automatic tensor conversion
+    auto input_records = inputs.view().records();
+    auto output_records = outputs.view().records();
+    cms::torch::alpaka::SoAMetadata<torchportable::ParticleSoA> inputs_metadata(batch_size);
+    inputs_metadata.append_block("features", input_records.pt(), input_records.eta(), input_records.phi());
+    cms::torch::alpaka::SoAMetadata<torchportable::RegressionSoA> outputs_metadata(batch_size);
+    outputs_metadata.append_block("preds", output_records.reco_pt());
+    cms::torch::alpaka::ModelMetadata<torchportable::ParticleSoA, torchportable::RegressionSoA> metadata(
+        inputs_metadata, outputs_metadata);
+
+    // inference
+    NvtxScopedRange move_to_device("Regression::move_to_device");
+    if (cms::torch::alpaka::device(event.queue()) != model_->device()) {
```

Comment:
@makortel would it make sense to implement a `beginStream` method for the alpaka `EDProducer`s that gives the correct device, or even a queue ?